File: windows/src/GoalMaker.App/ViewModels/SignInViewModel.cs
using CommunityToolkit.Mvvm.ComponentModel;
using CommunityToolkit.Mvvm.Input;
using GoalMaker.App.Localization;
using GoalMaker.Core.Account;
using GoalMaker.Core.Auth;

namespace GoalMaker.App.ViewModels;

/// <summary>
/// Drives the two-step email code sign-in. Success is observed through the auth session. A dev build
/// against the local stack is also handed a way to read the code out of the stack's own mailbox, and
/// fills it in itself, so a reset stack does not cost six digits of typing (docs/sign-in.md).
/// </summary>
public sealed partial class SignInViewModel : ObservableObject
{
    private readonly IAuthGateway auth;
    private readonly IStrings strings;
    private readonly Func<string, CancellationToken, Task<string?>>? devCode;

    // The mail lands within a second or so on a local stack; after this the owner types it.
    private static readonly TimeSpan Wait = TimeSpan.FromMilliseconds(600);
    private const int Attempts = 6;

    [ObservableProperty]
    [NotifyPropertyChangedFor(nameof(IsEmailStep))]
    private bool isCodeStep;

    [ObservableProperty]
    [NotifyPropertyChangedFor(nameof(CodeSentText))]
    private string email = string.Empty;

    [ObservableProperty]
    private string code = string.Empty;

    [ObservableProperty]
    [NotifyPropertyChangedFor(nameof(IsIdle))]
    private bool isBusy;

    [ObservableProperty]
    [NotifyPropertyChangedFor(nameof(HasError))]
    private string errorText = string.Empty;

    public SignInViewModel(
        IAuthGateway auth,
        IStrings strings,
        string? devBackend,
        Func<string, CancellationToken, Task<string?>>? devCode = null)
    {
        this.auth = auth;
        this.strings = strings;
        this.devCode = devCode;
        DevBackendText = devBackend is null ? null : strings.Get("SignIn.DevBackend", devBackend);
    }

    public bool IsEmailStep => !IsCodeStep;

    public bool IsIdle => !IsBusy;

    public bool HasError => ErrorText.Length > 0;

    public string CodeSentText => strings.Get("SignIn.CodeSent", Email.Trim());

    public string? DevBackendText { get; }

    /// <summary>The dev build can fetch the code itself, so the button to do it is worth showing.</summary>
    public bool HasDevCode => devCode is not null;

    [RelayCommand]
    private async Task SendCodeAsync()
    {
        var address = EmailAddress.Parse(Email);
        if (address is null)
        {
            ErrorText = strings.Get("SignIn.Error.InvalidEmail");
            return;
        }

        if (await RunAsync(() => auth.SendCodeAsync(address, CancellationToken.None)))
        {
            Code = string.Empty;
            IsCodeStep = true;
            await FillCodeAsync();
        }
    }

    /// <summary>
    /// Reads the code out of the local stack's mailbox and fills it in, which signs in, since a full
    /// code verifies itself. The mail takes a moment to land, so it asks a few times before giving up
    /// and leaving the owner to type it.
    /// </summary>
    [RelayCommand]
    private async Task FillCodeAsync()
    {
        if (devCode is null || IsBusy)
        {
            return;
        }

        var address = Email.Trim();
        for (var attempt = 0; attempt < Attempts && IsCodeStep && Code.Length == 0; attempt++)
        {
            if (attempt > 0)
            {
                await Task.Delay(Wait).ConfigureAwait(true);
            }

            if (await devCode(address, CancellationToken.None).ConfigureAwait(true) is { } code)
            {
                Code = code;
                return;
            }
        }
    }

    [RelayCommand]
    private async Task VerifyCodeAsync()
    {
        if (IsBusy)
        {
            return;
        }

        var address = EmailAddress.Parse(Email);
        if (address is null)
        {
            IsCodeStep = false;
            ErrorText = strings.Get("SignIn.Error.InvalidEmail");
            return;
        }

        var parsed = SignInCode.Parse(Code);
        if (parsed is null)
        {
            ErrorText = strings.Get("SignIn.Error.InvalidCode");
            return;
        }

        await RunAsync(() => auth.VerifyCodeAsync(address, parsed, CancellationToken.None));
    }

    [RelayCommand]
    private void UseAnotherEmail()
    {
        IsCodeStep = false;
        Code = string.Empty;
        ErrorText = string.Empty;
    }

    partial void OnEmailChanged(string value) => ErrorText = string.Empty;

    partial void OnCodeChanged(string value)
    {
        var digits = new string([.. value.Where(char.IsAsciiDigit).Take(SignInCode.Length)]);
        if (digits != value)
        {
            Code = digits;
            return;
        }

        ErrorText = string.Empty;
        if (digits.Length == SignInCode.Length && !IsBusy)
        {
            _ = VerifyCodeCommand.ExecuteAsync(null);
        }
    }

    private async Task<bool> RunAsync(Func<Task<AuthResult>> action)
    {
        IsBusy = true;
        ErrorText = string.Empty;
        try
        {
            var result = await action();
            switch (result)
            {
                case AuthResult.Success:
                    return true;
                case AuthResult.WrongOrExpiredCode:
                    Code = string.Empty;
                    ErrorText = strings.Get("SignIn.Error.WrongCode");
                    break;
                case AuthResult.TooManyRequests:
                    ErrorText = strings.Get("SignIn.Error.TooManyRequests");
                    break;
                case AuthResult.Offline:
                    ErrorText = strings.Get("SignIn.Error.Offline");
                    break;
                case AuthResult.Failed failed:
                    ErrorText = strings.Get("SignIn.Error.Other", failed.Detail);
                    break;
            }

            return false;
        }
        finally
        {
            IsBusy = false;
        }
    }
}

File: android/app/src/main/kotlin/com/goalmaker/app/ui/signin/SignInViewModel.kt
package com.goalmaker.app.ui.signin

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.goalmaker.app.application.auth.AuthGateway
import com.goalmaker.app.application.auth.AuthResult
import com.goalmaker.app.application.auth.DevSignIn
import com.goalmaker.app.domain.account.EmailAddress
import com.goalmaker.app.domain.account.SignInCode
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

/**
 * Drives the two-step email code sign-in. Success is observed through [AuthGateway.session]. A dev
 * build against the local stack reads the code out of the stack's own mailbox through [devCode] and
 * fills it in, and offers the dev account, which is that in one press (docs/sign-in.md).
 *
 * [onSignedIn] runs when a code is accepted, which is the owner proving who they are, so the app
 * lock has nothing left to ask. It is not called when a stored session comes back.
 */
class SignInViewModel(
    private val auth: AuthGateway,
    private val devCode: (suspend (String) -> String?)? = null,
    private val onSignedIn: () -> Unit = {},
) : ViewModel() {

    private val state = MutableStateFlow(SignInUiState(hasDevSignIn = devCode != null))
    val uiState: StateFlow<SignInUiState> = state.asStateFlow()

    fun onEmailChange(value: String) = state.update { it.copy(email = value, error = null) }

    fun onCodeChange(value: String) {
        val digits = value.filter(Char::isDigit).take(SignInCode.LENGTH)
        state.update { it.copy(code = digits, error = null) }
        if (digits.length == SignInCode.LENGTH) verifyCode()
    }

    fun sendCode() {
        val email = EmailAddress.parse(state.value.email)
            ?: return state.update { it.copy(error = SignInError.INVALID_EMAIL) }
        run(onSuccess = { it.copy(step = SignInStep.CODE, code = "") }, then = ::fillCode) { auth.sendCode(email) }
    }

    /**
     * Reads the code out of the local stack's mailbox and fills it in, which signs in, since a full
     * code verifies itself. The mail takes a moment to land, so it asks a few times before giving up
     * and leaving the owner to type it.
     */
    fun fillCode() {
        val fetch = devCode ?: return
        val email = state.value.email.trim()
        viewModelScope.launch {
            repeat(ATTEMPTS) { attempt ->
                val current = state.value
                if (current.step != SignInStep.CODE || current.code.isNotEmpty()) return@launch
                if (attempt > 0) delay(WAIT)
                fetch(email)?.let { code ->
                    onCodeChange(code)
                    return@launch
                }
            }
        }
    }

    fun verifyCode() {
        val current = state.value
        if (current.busy) return
        val email = EmailAddress.parse(current.email)
            ?: return state.update { it.copy(step = SignInStep.EMAIL, error = SignInError.INVALID_EMAIL) }
        val code = SignInCode.parse(current.code)
            ?: return state.update { it.copy(error = SignInError.INVALID_CODE) }
        run(onSuccess = { it }, then = onSignedIn) { auth.verifyCode(email, code) }
    }

    fun useAnotherEmail() = state.update { SignInUiState(email = it.email, hasDevSignIn = devCode != null) }

    /**
     * The dev account in one press: its code goes to the stack's mailbox like any other, and is read
     * back from there. It is an account of its own with its own data, so trying things out never
     * touches the one the owner signs in as.
     */
    fun signInAsDev() {
        if (devCode == null || state.value.busy) return
        state.update { it.copy(email = DevSignIn.EMAIL, error = null) }
        sendCode()
    }

    private fun run(
        onSuccess: (SignInUiState) -> SignInUiState,
        then: () -> Unit = {},
        action: suspend () -> AuthResult,
    ) {
        state.update { it.copy(busy = true, error = null) }
        viewModelScope.launch {
            val result = action()
            state.update { current ->
                val idle = current.copy(busy = false)
                when (result) {
                    AuthResult.Success -> onSuccess(idle)
                    AuthResult.WrongOrExpiredCode -> idle.copy(error = SignInError.WRONG_CODE, code = "")
                    AuthResult.TooManyRequests -> idle.copy(error = SignInError.TOO_MANY_REQUESTS)
                    AuthResult.Offline -> idle.copy(error = SignInError.OFFLINE)
                    is AuthResult.Failed -> idle.copy(error = SignInError.OTHER, errorDetail = result.detail)
                }
            }
            if (result == AuthResult.Success) then()
        }
    }

    private companion object {
        // The mail lands within a second or so on a local stack; after this the owner types it.
        const val WAIT = 600L
        const val ATTEMPTS = 6
    }
}
